MSAL/test/unit/native_auth/public/state_machine/mfa/AwaitingMFAStateTests.swift

```diff
@@ -0,0 +1,113 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+import XCTest
+@testable import MSAL
+
+final class AwaitingMFAStateTests: XCTestCase {
+
+    private var sut: AwaitingMFAState!
+    private var controller: MSALNativeAuthSignInControllerMock!
+    private var correlationId: UUID = UUID()
+
+    override func setUp() {
+        super.setUp()
+
+        controller = .init()
+        sut = .init(controller: controller, scopes: [], continuationToken: "continuationToken", correlationId: correlationId)
+    }
+
+    // MARK: - Delegates
+
+    // Request challenge
+
+    func test_requestChallenge_delegate_withError_shouldReturnCorrectError() {
+        let exp = expectation(description: "mfa state")
+
+        let expectedError = MFAError(type: .generalError, message: "test error", correlationId: correlationId)
+        let expectedState = MFARequiredState(controller: controller, scopes: [], continuationToken: "continuationToken", correlationId: correlationId)
+
+        let expectedResult: MFARequestChallengeResult = .error(
+            error: expectedError,
+            newState: expectedState
+        )
+        controller.requestChallengeResponse = .init(expectedResult, correlationId: correlationId)
+
+        let delegate = MFARequestChallengeDelegateSpy(expectation: exp, expectedError: expectedError)
+
+        sut.requestChallenge(delegate: delegate)
+        wait(for: [exp], timeout: 1)
+
+        XCTAssertNil(delegate.newMFARequiredState)
```

Comment:
Maybe you can add a bool variable to the `MFARequestChallengeDelegateSpy` that is set to `true` if the `onMFARequestChallengeError` is called, and check for it here. Just to make the test easier for the reader.

I mean something like `onMFARequestChallengeErrorWasCalled: Bool`

Reply:
In the MFAGetAuthMethodsDelegateSpy we specify how the expected error should be. Once this attribute is set, the `onMFAGetAuthMethodsError` method check that the returned error is the same as the expected one. If this error is not set and the error callback is called the test fail.
Please, see MFAGetAuthMethodsDelegateSpy line 103